### DISCUSSION TITLE
Support templating of sourceDirectory field

### DISCUSSION THREAD

**Discussion by emil-vdw** [renovatebot/renovate#28734] [Suggest an Idea]
### Tell us more.

We have multiple monorepos from which we build various private Python packages. These repos contain projects like so:
```
Monorepo
├── package-a
├── package-b
└── package-c
```
Each of these projects contain a `README.md`.

We would like to receive changelog links and expands on PRs opened by Renovate. Currently the only option is to create a `packageRule` for each project in the monorepo like so:
```
    ...
    {
      "customChangelogUrl": "https://github.com/CoolCompany/mono-repo",
      "sourceDirectory": "package-a",
      "matchDepNames": ["package-a"]
    },
    {
      "customChangelogUrl": "https://github.com/CoolCompany/mono-repo",
      "sourceDirectory": "package-b",
      "matchDepNames": ["package-b"]
    },
    ...
```

It would be really handy if we could template `sourceDirectory` and only need 1 package rule like so:
```
    {
      "customChangelogUrl": "https://github.com/CoolCompany/mono-repo",
      "sourceDirectory": "{{depName}}",
      "matchDepNames": ["package-a", "package-b", ...]
    },
```

I've read through the documentation and tested it and it seems that `sourceDirectory` does not support templating.

Is this a feature that you could add or would consider a contribution for?

**Comment by rarkins** [comment]
`sourceDirectory` is not currently a supported config option in the first place.

**Reply by emil-vdw** [reply]
First of all I want to be respectful of your and other maintainers' time so let me say I'm not an expert when with respect to Renovate.

I have seen it in [the documentation](https://docs.renovatebot.com/modules/datasource/custom/#usage) and have tested it and it seems to work in the non-template case (first JSON block above).
Should I start a discussion around supporting this as an "official" config option or is this something the maintainers do not want in the first place?

**Reply by rarkins** [reply]
That documentation is for customDatasources specifically. Are you using those?

**Reply by emil-vdw** [reply]
No, it's used exactly as in the example in the description, just in a `packageRules` entry (these internal packages are hosted in our private PyPI registry).

```json
"packageRules": [
    ...
    {
      "customChangelogUrl": "https://github.com/CoolCompany/mono-repo",
      "sourceDirectory": "package-a",
      "matchDepNames": ["package-a"]
    },
    {
      "customChangelogUrl": "https://github.com/CoolCompany/mono-repo",
      "sourceDirectory": "package-b",
      "matchDepNames": ["package-b"]
    },
    ...
]
```

**Reply by rarkins** [reply]
The documentation you referenced refers to its use within customDatasources, not packageRules.

**Reply by rarkins** [reply]
We restricted support for sourceUrl and sourceDirectory out of concern that it might interfere with caching or allow nuisance interference with caching of release notes. This needs to be looked into first

**Reply by emil-vdw** [reply]
I understand. Thanks for taking the time.

Is there anything further I can do from my side?

**Reply by rarkins** [reply]
@zharinov did you reach a conclusion about whether it's safe to allow users to configure sourceUrl and sourceDirectory in packageRules? i.e. our caching is based on package + sourceurl + sourcedirectory and not package alone

**Reply by zharinov** [reply]
Yes, I remember it got different cache keys for different `sourceUrl`. Haven't checked for directories, though.

**Reply by rarkins** [reply]
https://github.com/renovatebot/renovate/issues/28801

**Reply by emil-vdw** [reply]
Thank you very much for taking the time @rarkins :pray:.
